### PR TITLE
Runtime updates with version stamping

### PR DESCRIPTION
**Pull request by ssunny3141**

### CODE REVIEW

File: common/common.go [hunk 1]
```diff
@@ -123,6 +132,97 @@ func BuildFuncV20180708(verbose bool, fpath string, funcfile *FuncFileV20180708,
...
 }
 
+func imageStampFuncFile(fpath string, funcfile *FuncFile) (*FuncFile, error) {
+
+	dir := filepath.Dir(fpath)
+	dockerfile := filepath.Join(dir, "Dockerfile")
+
+	// detect if build and run image both are absent and runtime is not docker then update them
+	fmt.Println(funcfile.BuildImage)
+	fmt.Println(funcfile.RunImage)
```
Comment (roneetshaw) [inline common/common.go:142]: Are these print statements required?

Comment (ssunny3141) [reply common/common.go:142]: Will remove them

File: langs/python.go [hunk 2]
```diff
@@ -32,7 +34,14 @@ func (h *PythonLangHelper) GenerateBoilerplate(path string) error {
 		return err
 	}
 	depFile := "requirements.txt"
-	if err := ioutil.WriteFile(depFile, []byte(reqsPythonSrcBoilerplate), os.FileMode(0644)); err != nil {
+	fdkVersionInfo, err := h.GetLatestFDKVersion()
+	if err != nil {
+		fmt.Println("Unable to get latest FDK version, using default")
```
Comment (sudhaku) [inline langs/python.go:39]: Where are we setting fdkVersionInfo to default if err was not nil ?

File: common/common.go [hunk 3]
```diff
@@ -123,6 +132,95 @@ func BuildFuncV20180708(verbose bool, fpath string, funcfile *FuncFileV20180708,
 	return funcfile, nil
 }
 
+func imageStampFuncFile(fpath string, funcfile *FuncFile) (*FuncFile, error) {
+
+	dir := filepath.Dir(fpath)
+	dockerfile := filepath.Join(dir, "Dockerfile")
+
+	// detect if build and run image both are absent and runtime is not docker then update them
+	if !Exists(dockerfile) && funcfile.Runtime != FuncfileDockerRuntime && funcfile.BuildImage == "" && funcfile.RunImage == "" {
```
[the remark this answers is not in the document]
Comment (ssunny3141) [reply common/common.go:141]: This is to ensure that for runtime 'docker' we don't stamp versions.
fn init --runtime docker 
This check is there at other places too wherever dev/runtime image were getting picked.

File: langs/python.go (hunk 2 above)
Comment (ssunny3141) [reply langs/python.go:39]: @metamemelord

File: common/common.go [hunk 4]
```diff
@@ -123,6 +132,95 @@ func BuildFuncV20180708(verbose bool, fpath string, funcfile *FuncFileV20180708,
...
+	if !Exists(dockerfile) && funcfile.Runtime != FuncfileDockerRuntime && funcfile.Build_image == "" && funcfile.Run_image == "" {
+
+		helper := langs.GetLangHelper(funcfile.Runtime)
+
+		// get the lang without any version
+		langRuntime := helper.Runtime()
+
+		// if fallback is not supported then continue with same lang helper
+		if langs.IsFallbackSupported(langRuntime) {
+			helper = langs.GetFallbackLangHelper(langRuntime)
```
[the remark this answers is not in the document]
Comment (ssunny3141) [reply common/common.go:196]: Latest runtime if initialized using newer version (this version) of cli will always have build and runtime image flag set, so check on outer condition will fail and won't fallback.
This check is only func.yaml file where build and runtime images are not not.

File: common/common.go [hunk 5]
```diff
@@ -123,6 +132,95 @@ func BuildFuncV20180708(verbose bool, fpath string, funcfile *FuncFileV20180708,
...
+			return funcfile, err
+		}
+		funcfile.BuildImage = bi
+
+
+		// In case of multistage build there will be a runtime image
+		if helper.IsMultiStage() {
+			ri, err := helper.RunFromImage()
+			if err != nil {
+				return funcfile, nil
```
[the remark this answers is not in the document]
Comment (ssunny3141) [reply common/common.go:165]: Good catch! will change.

File: langs/python.go (hunk 2 above)
Comment (metamemelord) [reply langs/python.go:39]: We go via the current flow: assign `fdk` in requirements.txt and not add version. On the other hand, if we do get the latest version, we add `fdk>=<version>` in requirements.txt.

Comment (metamemelord) [reply langs/python.go:39]: Failure to fetch latest FDK version can mean one of two things:

1. Pypi is down, in this case user function will not deploy - this is sort unrecoverable at the moment.
2. Pypi changes response json. In this case, we have 2 choices: prevent user to deploy function altogether until we fix the parsing logic OR leave the fdk tag empty in requirements.txt and let pip pick latest automatically. Current flow takes the latter path.

**Review by metamemelord — CHANGES_REQUESTED** [review]
Please squash the commits. LGTM otherwise.

**Comment by roneetshaw** [comment]
Changes look good.

**Review by roneetshaw — APPROVED** [review]

File: common/funcfile.go [hunk 6]
```diff
@@ -234,6 +234,8 @@ func encodeFuncfileJSON(path string, ff *FuncFile) error {
 
 // EncodeFuncfileYAML encodes function file.
 func EncodeFuncfileYAML(path string, ff *FuncFile) error {
+	fmt.Println(ff.BuildImage)
+	fmt.Println(ff.RunImage)
```
[the remark this answers is not in the document]
Comment (ssunny3141) [reply common/funcfile.go]: Removed.

**Pull request merged**